src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -336,16 +476,108 @@ private BoundExpression RewriteStringConcatenationFourExprs(SyntaxNode syntax, B
+            }
+
+            return expr;
+        }
+
+        private BoundExpression WrapIntoReadOnlySpanOfCharIfNecessary(BoundExpression expr, MethodSymbol stringImplicitConversionToReadOnlySpan, MethodSymbol readOnlySpanCtorRefParamChar)
+        {
+            var exprType = expr.Type;
+
+            Debug.Assert(exprType is not null && (exprType.IsStringType() || exprType.IsCharType() || exprType.IsReadOnlySpanChar()));
+
+            if (expr.Type.IsReadOnlySpanChar())
+            {
+                return expr;
+            }
+            else if (exprType.IsStringType())
+            {
+                return BoundCall.Synthesized(expr.Syntax, receiverOpt: null, initialBindingReceiverIsSubjectToCloning: ThreeState.Unknown, stringImplicitConversionToReadOnlySpan, expr);
+            }
+            else
+            {
+                if (CanBePassedByReference(expr))
+                {
+                    // new ReadOnlySpan<char>(in <expr>)
+                    return new BoundObjectCreationExpression(
+                        expr.Syntax,
+                        readOnlySpanCtorRefParamChar,
+                        ImmutableArray.Create(expr),
+                        argumentNamesOpt: default,
+                        argumentRefKindsOpt: ImmutableArray.Create(RefKindExtensions.StrictIn),
+                        expanded: false,
+                        argsToParamsOpt: default,
+                        defaultArguments: default,
+                        constantValueOpt: null,
+                        initializerExpressionOpt: null,
+                        type: readOnlySpanCtorRefParamChar.ContainingType);
+                }
+                else
+                {
+                    var temp = _factory.StoreToTemp(expr, out var tempAssignment);
+
+                    var wrappedChar = new BoundObjectCreationExpression(
+                        expr.Syntax,
+                        readOnlySpanCtorRefParamChar,
+                        ImmutableArray.Create<BoundExpression>(temp),
+                        argumentNamesOpt: default,
+                        argumentRefKindsOpt: ImmutableArray.Create(RefKindExtensions.StrictIn),
+                        expanded: false,
+                        argsToParamsOpt: default,
+                        defaultArguments: default,
+                        constantValueOpt: null,
+                        initializerExpressionOpt: null,
+                        type: readOnlySpanCtorRefParamChar.ContainingType);
+
+                    // var temp = <expr>;
+                    // new ReadOnlySpan<char>(in temp)
+                    return new BoundSequence(
+                        expr.Syntax,
+                        ImmutableArray.Create(temp.LocalSymbol),
+                        ImmutableArray.Create<BoundExpression>(tempAssignment),
+                        wrappedChar,
+                        wrappedChar.Type);
```

Comment:
In other words, because this sequence only covers the single argument, the following argument is free to reuse the same register (and does, causing undefined behavior).